File: solutions/typescript/problems/[0323_Medium] Number of Connected Components in an Undirected Graph.ts
// @ts-ignore
class DisjointSet {
    private root: number[] = [];
    private rank: number[] = [];

    constructor(size: number) {
        for (let i = 0; i < size; i++) {
            this.root[i] = i;
            this.rank[i] = 1;
        }
    }

    find(x: number): number {
        if (x === this.root[x]) {
            return x;
        }
        return (this.root[x] = this.find(this.root[x]));
    }

    union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX !== rootY) {
            if (this.rank[x] > this.rank[y]) {
                this.root[rootY] = rootX;
            } else if (this.rank[x] < this.rank[y]) {
                this.root[rootX] = rootY;
            } else {
                this.root[rootY] = rootX;
                this.rank[rootX]++;
            }
        }
    }

    connected(x: number, y: number): boolean {
        return this.find(x) === this.find(y);
    }
}

function countComponents(n: number, edges: number[][]): number {
    const disjointSet = new DisjointSet(n);
    for (const [x, y] of edges) {
        disjointSet.union(x, y);
    }

    const components = new Set();
    for (let i = 0; i < n; i++) {
        components.add(disjointSet.find(i));
    }
    return components.size;
}

File: solutions/c++/problems/[0323_Medium] Number of Connected Components in an Undirected Graph.cpp
class DisjointSet {
    vector<int> root;
    vector<int> rank;
    unordered_set<int> activeRoots;
public:
    DisjointSet(int n) {
        root.reserve(n);
        rank = vector<int>(n, 0);
        for (int i = 0; i < n; i++) {
            root[i] = i;
            activeRoots.insert(i);
        }
    }

    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        activeRoots.insert(rootX);
        activeRoots.insert(rootY);
        if (rank[rootX] < rank[rootY]) {
            root[rootX] = rootY;
            activeRoots.erase(rootX);
        } else {
            root[rootY] = rootX;
            activeRoots.erase(rootY);
            if (rank[rootX] == rank[rootY]) rank[rootX]++;
        }
    }

    int find(int x) {
        if (root[x] == x) return x;
        return root[x] = find(root[x]);
    }

    bool connected(int x, int y) {
        return find(x) == find(y);
    }

    int getRootCount() {
        return activeRoots.size();
    }
};

class Solution {
public:
    int countComponents(int n, vector <vector<int>> &edges) {
        auto ds = new DisjointSet(n);
        for (auto edge: edges) if (!ds->connected(edge[0], edge[1])) ds->unionSet(edge[0], edge[1]);
        return ds->getRootCount();
    }
};